### ISSUE TITLE
Fine-tuning model with a common training loop

### ISSUE THREAD

**Issue by DT6A** [UKPLab/sentence-transformers#1299]
Hello

Is it possible to fine-tune model in a default way using model outputs? The schema I want to implement is the following:

```Python
model = SentenceTransformer('model-name')
# Freeze some layers
optimizer = torch.optim.Adam(model.parameters())
for inputs in data:
  embeddings = model(inputs)
  # Do some staff with embeddings and calculate loss
  loss.backward()
  optimizer.step()
```
Is it possible to implement and if so are there any examples for this? Training with `fit` function is quite a problem in my case and I need to fine-tune some parts of the pretrained model. More specifically I want to fine-tune embeddings layer only

**Comment by nreimers** [comment]
Yes, that is possible.

There are no examples directly. But if you look at the fit method, it does quite something like that.

You need to pass you input data through the tokenizer:
```
model = SentenceTransformer('model-name')
# Freeze some layers
optimizer = torch.optim.Adam(model.parameters())
for inputs in data:
  model_input = model.tokenize(inputs)
  model_output = model(model_input )
  # Do some staff with embeddings and calculate loss
  loss.backward()
  optimizer.step()
```

Now model_output contains a dict with various fields like token_embeddings, sentence_embeddings etc.

From there you can compute the loss

**Comment by DT6A** [comment]
> Yes, that is possible.
> 
> There are no examples directly. But if you look at the fit method, it does quite something like that.
> 
> You need to pass you input data through the tokenizer:
> 
> ```
> model = SentenceTransformer('model-name')
> # Freeze some layers
> optimizer = torch.optim.Adam(model.parameters())
> for inputs in data:
>   model_input = model.tokenize(inputs)
>   model_output = model(model_input )
>   # Do some staff with embeddings and calculate loss
>   loss.backward()
>   optimizer.step()
> ```
> 
> Now model_output contains a dict with various fields like token_embeddings, sentence_embeddings etc.
> 
> From there you can compute the loss

Got it. Thank you
